opentelemetry-api/src/opentelemetry/patcher/__init__.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# type: ignore
+
+"""
+OpenTelemetry patcher
+
+This includes the base patcher class and a no-op implementation.
+"""
+
+from abc import ABC, abstractmethod
+from functools import wraps
+from logging import getLogger
+
+_LOG = getLogger(__name__)
+
+
+class BasePatcher(ABC):
+    """An ABC for patchers"""
+
+    @staticmethod
+    def protect(class_) -> None:
```

Comment:
I think this name is confusing (or at least vague). Also, there is no need to use a decorator here. Just implement the "protection" logic in non-abstract patch/unpatch methods of the ABC and call some abstract `do_patch` and `do_unpatch` to do the actual work (https://en.wikipedia.org/wiki/Non-virtual_interface_pattern). Then we also can use a simple member variable `is_patch_active` instead of a somewhat arcane function attribute `protected`.

Reply:
You suggest a very accurate application of a pattern. I opted for the decorator approach to make it possible for the end user to decide to protect or not the `patch` and `unpatch` methods. After some thinking, I actually can't think of any good reason not to have this protection present always :man_shrugging:.

I usually don't deal with function attributes but I found it nice that they hid very well the protection mechanism from the rest of the class. Anyways, after taking all into consideration, I'll most likely refactor this along your suggestions.